kernel/src/engine/arrow_expression.rs

```diff
@@ -366,6 +366,62 @@ fn evaluate_expression(
     }
 }
 
+fn apply_schema(batch: RecordBatch, schema: &Schema) -> DeltaResult<RecordBatch> {
+    let cols = batch.columns();
+    let arrow_schema = batch.schema();
+    let sa = apply_to_cols(cols, None, arrow_schema.fields(), schema)?;
+    Ok(sa.into())
+}
+
+fn apply_to_cols(
+    cols: &[Arc<dyn Array>],
+    nulls: Option<NullBuffer>,
+    arrow_fields: &Fields,
+    schema: &Schema,
+) -> DeltaResult<StructArray> {
+    // build up a new set of (col, field) pairs
+    let result_iter = arrow_fields.into_iter().zip(schema.fields()).zip(cols).map(
+        |((arrow_field, kernel_field), col)| -> DeltaResult<(Arc<dyn Array>, ArrowField)> {
+            match (&kernel_field.data_type, arrow_field.data_type()) {
+                (DataType::Struct(kernel_fields), ArrowDataType::Struct(arrow_fields)) => {
+                    if kernel_fields.fields.len() != arrow_fields.len() {
+                        return Err(make_arrow_error(format!(
+                            "Kernel schema had {} fields, but data has {}",
+                            kernel_fields.fields.len(),
+                            arrow_fields.len()
+                        )));
+                    }
+                    let sa = col.as_struct_opt().ok_or(make_arrow_error(
+                        "Arrow claimed to be a struct but isn't a StructArray".to_string(),
+                    ))?;
+                    let (fields, sa_cols, sa_nulls) = sa.clone().into_parts();
+                    let transformed = apply_to_cols(&sa_cols, sa_nulls, &fields, kernel_fields)?;
+                    let new_field = arrow_field
+                        .as_ref()
+                        .clone()
+                        .with_name(kernel_field.name.clone())
+                        .with_data_type(transformed.data_type().clone());
+                    Ok((Arc::new(transformed) as Arc<dyn Array>, new_field))
+                }
+                _ => {
+                    ensure_data_types(&kernel_field.data_type, arrow_field.data_type())?;
+                    Ok((
+                        col.clone(),
+                        arrow_field
+                            .as_ref()
+                            .clone()
+                            .with_name(kernel_field.name.clone()),
+                    ))
+                }
+            }
+        },
+    );
+    let (new_cols, new_fields): (Vec<Arc<dyn Array>>, Vec<ArrowField>) =
+        result_iter.process_results(|iter| iter.unzip())?;
```

Comment:
Yep. I basically think of it as, if i have a function that takes `Iter<T>`, and I want to apply it over a iterator of `Result<T, E>`, then I can call `process_results` with the function.

`process_results` takes care of unwrapping the `Ok` values, calling the process function on  all the inner values of the `Ok`s, and then wraps the final result in an `Ok`. Otherwise it just stops at the first `Err`. 

> I think this is functionally equivalent to collect .. iter .. unzip .. collect?

Almost, without the final collect, since unzip is basically two collects in a trench-coat.

Reply:
TIL!